extensions/ql-vscode/src/query-history.ts

```diff
@@ -162,120 +168,284 @@ export class QueryHistoryManager {
-    this.treeDataProvider.remove(queryHistoryItem);
-    queryHistoryItem.dispose();
+  async handleRemoveHistoryItem(
+    singleItem: CompletedQuery,
+    multiSelect: CompletedQuery[]
+  ) {
+    (multiSelect || [singleItem]).forEach((item) => {
+      this.treeDataProvider.remove(item);
+      item.dispose();
+    });
     const current = this.treeDataProvider.getCurrent();
     if (current !== undefined) {
       this.treeView.reveal(current);
       await this.invokeCallbackOn(current);
     }
   }
 
-  async handleSetLabel(queryHistoryItem: CompletedQuery) {
+  async handleSetLabel(
+    singleItem: CompletedQuery,
+    multiSelect: CompletedQuery[]
+  ): Promise<void> {
+    if (!this.assertSingleQuery(multiSelect)) {
+      return;
+    }
+
     const response = await vscode.window.showInputBox({
       prompt: 'Label:',
       placeHolder: '(use default)',
-      value: queryHistoryItem.getLabel(),
+      value: singleItem.getLabel(),
     });
     // undefined response means the user cancelled the dialog; don't change anything
     if (response !== undefined) {
       if (response === '')
-        // Interpret empty string response as "go back to using default"
-        queryHistoryItem.options.label = undefined;
-      else
-        queryHistoryItem.options.label = response;
+        // Interpret empty string response as 'go back to using default'
+        singleItem.options.label = undefined;
+      else singleItem.options.label = response;
       this.treeDataProvider.refresh();
     }
   }
 
-  async handleCompareWith(query: CompletedQuery) {
-    const dbName = query.database.name;
-    const comparableQueryLabels = this.treeDataProvider.allHistory
-      .filter((otherQuery) => otherQuery !== query && otherQuery.didRunSuccessfully && otherQuery.database.name === dbName)
-      .map(otherQuery => ({
-        label: otherQuery.toString(),
-        description: otherQuery.databaseName,
-        detail: otherQuery.statusString,
-        query: otherQuery
-      }));
-    const choice = await vscode.window.showQuickPick(comparableQueryLabels);
-    if (choice) {
-      this.doCompareCallback(query, choice.query);
+  async handleCompareWith(
+    singleItem: CompletedQuery,
+    multiSelect: CompletedQuery[]
+  ) {
+    try {
+      if (!singleItem.didRunSuccessfully) {
+        throw new Error('Please select a successful query.');
+      }
+
+      const from = singleItem;
+      const to = await this.findOtherQueryToCompare(singleItem, multiSelect);
+
+      if (from && to) {
+        this.doCompareCallback(from, to);
+      }
+    } catch (e) {
+      helpers.showAndLogErrorMessage(e.message);
     }
   }
 
-  async handleItemClicked(queryHistoryItem: CompletedQuery) {
-    this.treeDataProvider.setCurrentItem(queryHistoryItem);
+  async handleItemClicked(
+    singleItem: CompletedQuery,
+    multiSelect: CompletedQuery[]
+  ) {
+    if (!this.assertSingleQuery(multiSelect)) {
```

Comment:
I'm not sure if this requires any change to your code here, but I observe that the `command` of a `TreeItem` seems to only be invoked when the selection is a single item, so I think this assertion will never fail. Maybe log an additional invariant violation message if it does?

Reply:
I'm not sure what you mean here? The behaviour is that when there is a single selection, the command is invoked with `singleItem` being the one that was clicked on and `multiSelect` being undefined.

When there are multiple selections, `singleItem` is still the one being clicked on and `multiSelect` contains the list of items selected.

Reply:
Correct. I think (but I'm not 100% confident) the only time vscode will ever invoke this particular event handler, it will be the case that `multiSelect === undefined`.